src/model.jl

```diff
@@ -104,100 +98,179 @@ implementation simply uses `T.name.wrapper` which is the `UnionAll` type corresp
-_keys(params::Tuple{}, m::AbstractModel) = ()
-
-@inline function Base.getindex(m::AbstractModel, nm::Symbol)
-    if nm == :component
-        paramparenttypes(m)
-    elseif nm == :fieldname
-        paramfieldnames(m)
+# Indexing interface
+@inline Base.getindex(m::AbstractModel, col::Symbol) = getindex(m, :, col)
+@inline Base.getindex(m::AbstractModel, i) = getindex(m, i, :)
+@inline Base.getindex(m::AbstractModel, ::Colon, ::Colon) = m
+@inline function Base.getindex(m::AbstractModel, i, col)
+    return if col == :component
+        paramcomponents(m)[i]
+    elseif col == :fieldname
+        paramfieldnames(m)[i]
     else
-        map(p -> getindex(p, nm), params(m))
+        _getindex(params(m), i, col)
     end
 end
-
-function Base.show(io::IO, mime::MIME"text/plain", m::AbstractModel)
-    show(io, mime, typeof(m))
-    println(io, " with parent object of type: \n")
-    show(io, mime, typeof(parent(m)))
-    println(io, "\n\n")
-    printparams(io::IO, m)
+@inline Base.setindex(m::AbstractModel, xs, col::Union{Symbol,Type{<:Val}}) = Base.setindex(m, xs, :, col)
+@inline Base.setindex(m::AbstractModel, xs, i) = Base.setindex(m, xs, i, :)
+@inline Base.setindex(m::AbstractModel, xs, i, col::Symbol) = Base.setindex(m, xs, i, Val{col})
+@inline Base.setindex(m::AbstractModel, xs, i, ::Type{Val{col}}) where col = Base.setindex(m, xs, collect(i), Val{col})
+@inline Base.setindex(m::AbstractModel, xs, i, ::Colon) = Base.setindex(m, xs, collect(i), filter(!_isreserved, keys(m)))
+@inline Base.setindex(m::AbstractModel, xs, i::Integer, ::Colon) = Base.setindex(m, xs, i, filter(!_isreserved, keys(m)))
+@inline Base.setindex(m::AbstractModel, xs, i::Integer, ::Type{Val{col}}) where col = _setindex(m, xs, i, Val{col})
+@inline function Base.setindex(m::AbstractModel, xs, i, cols::Union{Tuple,AbstractVector})
+    for col in cols
+        m = Base.setindex(m, Tables.getcolumn(xs, col), i, Val{col})
+    end
+    return m
 end
-
-printparams(m) = printparams(stdout, m)
-function printparams(io::IO, m::AbstractModel)
-    if length(m) > 0
-        println(io, "Parameters:")
-        PrettyTables.pretty_table(io, m, [keys(m)...])
+@inline function Base.setindex(m::AbstractModel, xs, i::AbstractVector, ::Type{Val{col}}) where col
+    @assert !_isreserved(col) "column name :$col is reserved and cannot be modified"
+    @assert col ∈ keys(m) "column $col does not exist"
+    return _setindex(m, xs, i, Val{col})
+end
+@inline function Base.setindex(m::AbstractModel, xs, ::Colon, ::Type{Val{col}}) where col
+    @assert !_isreserved(col) "column name :$col is reserved and cannot be modified"
+    return if col ∈ keys(m)
+        _setindex(m, xs, :, Val{col})
+    else
+        _addindex(m, xs, Val{col})
     end
 end
+@inline Base.setindex!(m::AbstractModel, xs, col::Union{Symbol,Type{Val}}) = setindex!(m, xs, :, col)
+@inline Base.setindex!(m::AbstractModel, xs, i) = setindex!(m, xs, i, :)
```

Comment:
It makes me a little uneasy guessing row/col like this based on type. Can you do this with any other tables?

And should we maybe make `i::Int` ?